README.md
# MCO Jelly Noodles
>MCO Jelly Noodles is a script to deal with the issue of inefficiency in the presentation of the booking table data for our facility managers working with the MCOnline Facility booking system.

## Table of Contents
- [Current Version](#current-version)
- [Installation](#installation)
  - [Command Menu](#command-menu)
- [Installation(Alternative)](#alternative-installation)
- [Usage](#usage)
  - [Usage(Alternative)](#alternative-usage)
  - [Demonstration](#demonstration)
- [Roadmap](#roadmap)
- [Contributing](#contributing)


## Current version: 
  [3.0.7]
##  Installation
<details><summary>Disclaimer</summary><p>As of now, the only 'permanent' way to install the script is to save the script directly into the browser.
Currently the script can only be 'installed' in chrome(as far as we know), without implementing any third-party software or browser extensions. Testing has not been done on any other browsers.</p></details>

<details><summary>Possible future plans</summary><p>
we might expand the project into a userscript, in which case this section of the documentation will be updated. For now the only way to 'install' this script is via chrome's snippet devtool. running the script in any browser's console works fine(except IE9 and below). Saving the script as a bookmark also works.
</p></details>


### Command menu

In chrome, open up the [command menu.](https://developers.google.com/web/tools/chrome-devtools/command-menu/)

Type in `Snippet` and click the first option.

Paste in the [source code](#current-version) for the script.

Run the script using Ctrl + Enter, or rightclick and run.

## Alternative Installation
Alternatively, copy the code from index.js and save it as the URL of a bookmark on your browser. Anytime you need to use it just run.

## Usage

To use the script, the snippet has to be run.
>Ctrl + shift + I  -->  Ctrl + shift + P  --> backspace, type: "!"  -->  press enter.

### Alternative Usage

Run the bookmark that contains the script code.

### Demonstration

Upon loading the script, two buttons will appear above the calendar. 

<img src="demo/jifwan 1.gif" alt="figure 1">

| Button | Description |
| --- | ---- |
| Pick a Date | The First button, labelled "Pick a Date", has to be used first, to initialize the date used in the script. The "Bookings" button will not work otherwise. Upon clicking the button, an alert window will pop up on the page, asking for a date to be typed into the input box. Any date can be given, however only the date that is within the week that is shown in the calendar will be taken for the "Bookings" buton to work. |
| Bookings | The Second button, labelled "Bookings", can be used at any point in time as long as the date is initialized using the "Pick a Date" button as mentioned above. The "Bookings" button, upon being clicked, will replace the entire calendar with the information of all the bookings on the date specified. |

<img src="demo/jifwan 2.gif" alt="figure 2">

## Roadmap
Future developments will still be continued unless otherwise specified.

>Planned features:
```bash
-Make the design prettier?
```

## Contributing
Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

Please make sure to update tests as appropriate.


[3.0.7]: index.js



Changelog.md
# Changelog
All notable changes to this project will be documented in this file.

## [Version 3.0.7 - 27/8/2019 (current)]
###Fixed
-alert error when picking date.
>>upon picking a date that was in a week that included days of the next month, an alert pops up preventing the selection of date.

## Version 3.0.6 - 6/8/2019
###Minor feature implementations:
>>Toggle between calendar and bookings display using the Bookings button.

## Version 3.0.5 - 5/8/2019
###Fixed
-system not accepting proper date values.
>>upon picking a date that was in a week that is supposed to be valid, an alert pops up saying otherwise.

## Version 3.0.4 - 1/8/2019
###Fixed
-alert error when picking date.
>>upon picking a date that was in a week that included days of the next month, an alert pops up preventing the selection of date.

## Version 3.0.3
###Fixed
-Retention of entire calendar when changing 'Facilitiy Name' dropdown.
>>upon switching facilities using the filters on the left-side container, and then clicking on "Bookings", calendar was changed to the calendar values that were from the previous facility booking calendar.
>>This issue existed whether or not the calendar was displayed as a calendar or data.

-Issue of adding too many buttons when running the script multiple times.
>>running the script more than one time will add another set of buttons each time the script is run.

## Version 3.0.2
###Minor feature implementations:
>>Toggle between calendar and bookings display using the Bookings button.

## Version 3.0.1
###Added
>>Initialize today's current date as the date upon running the script.

## Version 3
###Changed 
-Style and design of the displayed data.
###Added
>>Error sequence when trying to display bookings when there is no date specified.
>>Error sequence when trying to input a date not within the currently displayed week.
>>Error sequence when there is no bookings for the date specified.

## Version 2.1.14
###Fixed
-Retention of output after switching 'Facility Name' dropdown.
>>upon switching facilities using the filters on the left-side container, and then clicking on "Bookings", the results showed a significant number of data results containing mostly 'undefined'. 
>>upon further repetition of the above, the quantity of the 'undefined' results increased.

## Version 2
###Added
-Buttons.
>Offloaded the functionality of the script into two buttons that appear on the page upon loading the script onto the website.
###Fixed
-Issue where script needed to be copied and pasted multiple times for one job.

## Version 1
>Let there be light.

[Version 3.0.7 - 27/8/2019 (current)]: src/source.js


index.js
javascript:var x,q,state,y=[],z=[],cal=[],date=(new Date).toLocaleDateString(),st=(new Date).toLocaleDateString("en-GB",{day:"numeric",month:"short",year:"numeric"});function reinitializedate(){if(x=document.getElementsByClassName("order-booking"),null!=(date=prompt("Date(DD/MM/YYYY):"))){date.includes("/")?date=date.split("/")[1]+"/"+date.split("/")[0]+"/"+date.split("/")[2]:date.includes(" ")&&(date=date.split(" ")[0]+" "+date.split(" ")[1]+" "+date.split(" ")[2]),st=new Date(date).toLocaleDateString("en-GB",{day:"numeric",month:"short",year:"numeric"});var t=document.getElementById("txtdatetimeshow").textContent.split(" - "),e=parseInt(t[0][0]+t[0][1]),a=parseInt(t[1][0]+t[1][1]),i=t[0].split(" ")[1],n=t[1].split(" ")[1],r=st.split(" ")[1],d=parseInt(st.split(" ")[0][0]+st.split(" ")[0][1]);i!=n&&d>a&&d<e?alert("date selected not within the current week."):i==n&&(d<e||d>a)?alert("date selected not within the current week."):r==i||r==n||alert("date selected not within the current week.")}else alert("no date selected.")}function reformatdisplay(){if(0==document.getElementsByClassName("calendar-week-view").length&&document.getElementsByClassName("BookingCard").length>0&&(q=!0),document.getElementsByClassName("calendar-week-view").length>0&&(q=!1),q)q&&(document.getElementsByClassName("calendar")[0].innerHTML=cal,q=!1);else{cal=document.getElementsByClassName("calendar")[0].innerHTML,x=document.getElementsByClassName("order-booking");var t=0;if(null==date)return void alert("no date selected.");for(i=0;i<x.length;i++)x[i].title.includes(st)&&(y.push(x[i].title),t++);if(0==t&&0==q)return void alert("No bookings for the date selected.");for(i=0;i<y.length;i++)y[i].includes("Subject:")&&(y[i]=y[i].split("Subject:")[0]),y[i].includes("Status: Auto approved")&&(y[i]=y[i].split("Status: Auto approved")[0]+y[i].split("Status: Auto approved")[1]);for(i=0;i<y.length;i++)y[i]=st+" "+y[i].split(st)[1],z[0]=document.createElement("div"),z[0].setAttribute("class","BookingCard"),z[0].setAttribute("style","max-width:80%;height:auto;margin:20px auto; background: cyan; padding: 20px; overflow: hidden; border-radius: 8px; box-shadow: 5px 10px; transition: 0.6s; background-image: linear-gradient(to right, #FFFFFF, #0652DD, #12CBC4);"),z[1]=document.createElement("h1"),z[1].setAttribute("style","margin:0 0; padding: 0;"),z[1].textContent=y[i].split(st)[1].split("Remark: ")[0],z[2]=document.createElement("h3"),z[2].setAttribute("style","margin:0 0; padding: 0;"),z[2].textContent="Remark: "+y[i].split("Remark: ")[1],z[0].appendChild(z[1]),z[0].appendChild(z[2]),z[0].append(st),y[i]=z[0],z=[];var e=document.getElementsByClassName("calendar");for(e[0].innerHTML="",i=0;i<y.length;i++)e[0].appendChild(y[i]);e=[],y=[],z=[],t=0,q=!0}document.getElementById("inputFacilityName").focus()}var cheytan=document.createElement("div");cheytan.setAttribute("style","float:right;width: 50%;height:35px;margin-top: 0;");var rdb=document.createElement("div");rdb.setAttribute("class","fbutton");var rid=document.createElement("button");rid.setAttribute("class","button2 btn"),rid.setAttribute("onclick","reinitializedate()"),rid.setAttribute("id","dapeTicker"),rid.innerHTML="Pick a Date",rdb.appendChild(rid);var srb=document.createElement("div");srb.setAttribute("class","fbutton");var rfd=document.createElement("button");rfd.setAttribute("class","button2 btn"),rfd.setAttribute("onclick","reformatdisplay()"),rfd.setAttribute("id","quantumFacibrator"),rfd.innerHTML="Bookings",srb.appendChild(rfd),cheytan.appendChild(rdb),cheytan.appendChild(srb),null!=state&&0!=state||(document.getElementById("caltoolbar").querySelector(".clear").parentNode.insertBefore(cheytan,document.getElementById("caltoolbar").firstElementChild.lastElementChild),state=1);


src/source.js
//Initialize global variables.
var x;
var y = [];
var z = [];
var q;
var cal = [];
var state;

//Initialize current date.
var date = new Date().toLocaleDateString();
var st = new Date().toLocaleDateString('en-GB', {
    day: 'numeric',
    month: 'short',
    year: 'numeric'
});

//Initialize method for selecting the date.
function reinitializedate() {
    x = document.getElementsByClassName("order-booking");
    date = prompt("Date(DD/MM/YYYY):");
    //when user hits cancel.
    if (date == null) {
        alert("no date selected.");
	//console.log("User hit cancel.");
        return;
    }
    //parse the input date into a format recognizable by the script.
    if (date.includes('/')) {
        date = (date.split("/")[1] + "/" + date.split("/")[0] + "/" + date.split("/")[2]);
    } else if (date.includes(' ')) {
        date = (date.split(" ")[0] + " " + date.split(" ")[1] + " " + date.split(" ")[2]);
    }
    //parse input date into format used in SG
    st = new Date(date).toLocaleDateString('en-GB', {
        day: 'numeric',
        month: 'short',
        year: 'numeric'
    });console.log(`Parsed date = ${st}.`);
    //sample = 05 Aug 2019 - 11 Aug 2019 || 27 Aug 2019 - 03 Sep 2019
    var lookatTimeOfWeek = document.getElementById("txtdatetimeshow").textContent.split(" - ");
    var firstDay = parseInt(lookatTimeOfWeek[0][0] + lookatTimeOfWeek[0][1]);
    var lastDay = parseInt(lookatTimeOfWeek[1][0] + lookatTimeOfWeek[1][1]);
    var firstMonth = lookatTimeOfWeek[0].split(" ")[1];
    var lastMonth = lookatTimeOfWeek[1].split(" ")[1];
    var currentMonth = st.split(" ")[1];
    var currentDay = parseInt(st.split(" ")[0][0] + st.split(" ")[0][1]);
    //comparison check to ensure the input date is within the week shown in calendar range.
      //this checks if end of week is in the next month.
    if (firstMonth != lastMonth) {
        if (currentDay > lastDay && currentDay < firstDay) {
            alert("date selected not within the current week.");
	    console.log(`${currentDay} > ${lastDay} && ${currentDay} < ${firstDay}`);
            return;
        }
    }
      //this checks if end of week is within current month.
    if (firstMonth == lastMonth) {
        if (currentDay < firstDay || currentDay > lastDay) {
            alert("date selected not within the current week.");
	    console.log(`${currentDay} > ${firstDay} || ${currentDay} < ${lastDay}`);
            return;
        }
    }
      //this checks if the input month was even in the calendar range.
    if (currentMonth != firstMonth && currentMonth != lastMonth) {
      alert("date selected not within the current week.");
      console.log(`${currentMonth} != ${firstMonth} && ${currentMonth} != ${lastMonth}`);
      return;
    }
}

//Initialize method for displaying the data.
function reformatdisplay() {
    if (document.getElementsByClassName('calendar-week-view').length == 0 && document.getElementsByClassName('BookingCard').length > 0) {
        q = true;
    }
    if (document.getElementsByClassName('calendar-week-view').length > 0) {
        q = false;
    }
    if (!q) {
        cal = document.getElementsByClassName("calendar")[0].innerHTML;
        x = document.getElementsByClassName("order-booking");
        var timings = 0;
        if (date == null) {
            alert("no date selected.");
            return;
        }
        for (i = 0; i < x.length; i++) {
            if (x[i].title.includes(st)) {
                y.push(x[i].title);
                timings++;
            }
        }
        if (timings == 0 && q == false) {
            alert('No bookings for the date selected.');
            return;
        }
        for (i = 0; i < y.length; i++) {
            if (y[i].includes("Subject:")) {
                y[i] = y[i].split("Subject:")[0];
            }
            if (y[i].includes("Status: Auto approved")) {
                y[i] = y[i].split("Status: Auto approved")[0] + y[i].split("Status: Auto approved")[1];
            }
        }
        for (i = 0; i < y.length; i++) {
            y[i] = st + " " + y[i].split(st)[1];
            z[0] = document.createElement('div');
            z[0].setAttribute('class', "BookingCard");
            z[0].setAttribute('style', "max-width:80%;height:auto;margin:20px auto; background: cyan; padding: 20px; overflow: hidden; border-radius: 8px; box-shadow: 5px 10px; transition: 0.6s; background-image: linear-gradient(to right, #FFFFFF, #0652DD, #12CBC4);");
            z[1] = document.createElement('h1');
            z[1].setAttribute('style', "margin:0 0; padding: 0;");
            z[1].textContent = y[i].split(st)[1].split("Remark: ")[0];
            z[2] = document.createElement('h3');
            z[2].setAttribute('style', "margin:0 0; padding: 0;");
            z[2].textContent = "Remark: " + y[i].split("Remark: ")[1];
            z[0].appendChild(z[1]);
            z[0].appendChild(z[2]);
            z[0].append(st);
            y[i] = z[0];
            z = [];
        }
        var rewrite = document.getElementsByClassName("calendar");
        rewrite[0].innerHTML = "";
        for (i = 0; i < y.length; i++) {
            rewrite[0].appendChild(y[i]);
        }
        rewrite = [], y = [], z = [], timings = 0;
        q = true;
    } else if (q) {
        document.getElementsByClassName("calendar")[0].innerHTML = cal;
        q = false;
    }
    document.getElementById("inputFacilityName").focus()
}

//Create a new element to contain the buttons.
var cheytan = document.createElement('div');
cheytan.setAttribute('style', "float:right;width: 50%;height:35px;margin-top: 0;")

//Button Creation: Date Select.
var rdb = document.createElement('div');
rdb.setAttribute('class', "fbutton");

var rid = document.createElement('button');
rid.setAttribute('class', "button2 btn");
rid.setAttribute('onclick', "reinitializedate()");
rid.setAttribute('id', "dapeTicker");
rid.innerHTML = "Pick a Date";

rdb.appendChild(rid);


//Button Creation: Data Display.
var srb = document.createElement('div');
srb.setAttribute('class', "fbutton");

var rfd = document.createElement('button');
rfd.setAttribute('class', "button2 btn");
rfd.setAttribute('onclick', "reformatdisplay()");
rfd.setAttribute('id', "quantumFacibrator");
rfd.innerHTML = "Bookings";

srb.appendChild(rfd);


//Insert buttons to container created at the start(after initialization).
cheytan.appendChild(rdb);
cheytan.appendChild(srb);
//TODO: Debug css

//Add the whole thing to the website live, conditionally.
if (state == null || state == 0) {
    document.getElementById("caltoolbar").querySelector('.clear').parentNode.insertBefore(cheytan, document.getElementById("caltoolbar").firstElementChild.lastElementChild);
    state = 1;
}
